### ISSUE TITLE
`invalid_value` does not warn when transmuting from a unit struct to an uninhabited type

### ISSUE THREAD

**Issue by WaffleLapkin** [rust-lang/rust#126529]
I tried this code:

```rust
struct T;
enum U {}

pub fn ub() {
    unsafe { std::mem::transmute::<T, U>(T) };
}

fn main() {}
```

I expected to see this happen: warning that the transmute is invalid/UB.

Instead, this happened: no warning :(

A similar example, but with `()` instead of a unit structure _does_ warn:

```rust
pub fn ub() {
    unsafe { std::mem::transmute::<(), U>(()) };
}

fn main() {}
```
```text
warning: the type `U` does not permit zero-initialization
  --> tests/ui/never_type/never_transmute_never.rs:12:14
   |
12 |     unsafe { std::mem::transmute::<(), U>(()) };
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ this code causes undefined behavior when executed
   |
note: enums with no inhabited variants have no valid value
  --> tests/ui/never_type/never_transmute_never.rs:9:1
   |
9  | enum U {}
   | ^^^^^^
   = note: `#[warn(invalid_value)]` on by default
```

### Meta

`rustc --version --verbose`:
```
rustc 1.80.0-nightly (98489f248 2024-06-06)
binary: rustc
commit-hash: 98489f2487465f3765e5dd28d7305ebfd40f0865
commit-date: 2024-06-06
host: x86_64-unknown-linux-gnu
release: 1.80.0-nightly
LLVM version: 18.1.7
```
